cmd/src/lsif_upload.go

```diff
@@ -331,69 +202,36 @@ type errorWithHint struct {
 }
 
 func (e errorWithHint) Error() string {
-	return fmt.Sprintf("error: %s\n\n%s\n", e.err, e.hint)
-}
-
-type lsifUploadVerbosity int
-
-const (
-	lsifUploadVerbosityNone                  lsifUploadVerbosity = iota // -trace=0 (default)
-	lsifUploadVerbosityTrace                                            // -trace=1
-	lsifUploadVerbosityTraceShowHeaders                                 // -trace=2
-	lsifUploadVerbosityTraceShowResponseBody                            // -trace=3
-)
-
-type lsifUploadRequestLogger struct {
-	verbosity lsifUploadVerbosity
-}
-
-func (l *lsifUploadRequestLogger) LogRequest(req *http.Request) {
-	if l.verbosity == lsifUploadVerbosityNone {
-		return
-	}
-
-	if l.verbosity >= lsifUploadVerbosityTrace {
-		fmt.Printf("> %s %s\n", req.Method, req.URL)
-	}
-
-	if l.verbosity >= lsifUploadVerbosityTraceShowHeaders {
-		fmt.Printf("> Request Headers:\n")
-		for _, k := range sortHeaders(req.Header) {
-			fmt.Printf(">     %s: %s\n", k, req.Header[k])
-		}
-	}
-
-	fmt.Printf("\n")
+	return fmt.Sprintf("%s\n\n%s\n", e.err, e.hint)
 }
 
-func (l *lsifUploadRequestLogger) LogResponse(req *http.Request, resp *http.Response, body []byte, elapsed time.Duration) {
-	if l.verbosity == lsifUploadVerbosityNone {
-		return
-	}
-
-	if l.verbosity >= lsifUploadVerbosityTrace {
-		fmt.Printf("< %s %s %s in %s\n", req.Method, req.URL, resp.Status, elapsed)
+var errUnauthorizedHint = strings.Join([]string{
+	"You may need to specify or update your GitHub access token to use this endpoint.",
+	"See https://docs.sourcegraph.com/cli/references/lsif/upload.",
+}, "\n")
+
+// handleLSIFUploadError writes the given error to the given output and then
+// exits the process. If the given output object is nil then the error will
+// be written to standard out.
+func handleLSIFUploadError(out *output.Output, err error) {
+	if out == nil {
+		out = emergencyOutput()
 	}
 
-	if l.verbosity >= lsifUploadVerbosityTraceShowHeaders {
-		fmt.Printf("< Response Headers:\n")
-		for _, k := range sortHeaders(resp.Header) {
-			fmt.Printf("<     %s: %s\n", k, resp.Header[k])
-		}
+	if err == upload.ErrUnauthorized {
+		err = errorWithHint{err: err, hint: errUnauthorizedHint}
 	}
+	out.WriteLine(output.Linef(output.EmojiFailure, output.StyleWarning, "error: %s", err))
 
-	if l.verbosity >= lsifUploadVerbosityTraceShowResponseBody {
-		fmt.Printf("< Response Body: %s\n", body)
+	if !lsifUploadFlags.ignoreUploadFailures {
+		// Report but don't return the error
+		os.Exit(0)
```

Comment:
I think it is a bit weird if the command somewhere deep down calls os.Exit itself and doesn't let the command executor decide what should happen when a command returns an error, wdyt?

Reply:
We could early-return nil here instead - would that make sense? Do you have the same feelings on the os.Exit(1)?

Reply:
Yeah, sorry I meant all calls to `os.Exit` outside of the command handler. Not a blocker, but I'm wondering if we could just pass up the error, so in the future we might be able to have things like an error reporter in the command executor or something like that.

Reply:
Yeah that makes sense, will convert to a non-terminal flow here.